File: Andriod/Constructor/construction/src/main/java/com/palitri/openiot/construction/framework/web/api/OpenIotService.java
package com.palitri.openiot.construction.framework.web.api;

import com.palitri.openiot.construction.framework.web.models.Device;
import com.palitri.openiot.construction.framework.web.models.PresetsCollection;
import com.palitri.openiot.construction.framework.web.models.Project;
import com.palitri.openiot.construction.framework.web.models.Preset;
import com.palitri.openiot.construction.framework.web.networking.WebRequest;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public class OpenIotService {

    private String token;

    private String baseUrl;

    public OpenIotService()
    {
        this.baseUrl = "http://api.iot.cafe";
    }

    public OpenIotService setToken(String token)
    {
        this.token = token;

        return this;
    }

    public void requestUserLogin(String email, String password, Object... params)
    {
        new WebRequest(){
            @Override
            public void onResponse(String response, Object... params) {
                super.onResponse(response, params);

                OpenIotService service = (OpenIotService)params[1];
                service.userLoginResponse(response, params);
            }
        }
                .setRequestBody("{\"email\":\"" + email + "\",\"password\":\"" + password + "\"}")
                .setHttpMethod("POST")
                .send(this.baseUrl + "/users", this);
    }

    private void userLoginResponse(String response, Object... params)
    {
        try{
            if (response == null)
                this.token = null;
            else {
                JSONObject jsonResponse = new JSONObject(response);

                this.token = jsonResponse.getString("token");
            }

            this.onUserLoginResponse(this.token, params);

        } catch (JSONException e) {
            e.printStackTrace();
        }


    }

    public void onUserLoginResponse(String token, Object... params)
    {

    }



    public void requestUserProjects()
    {
        new WebRequest(){
            @Override
            public void onResponse(String response, Object... params) {
                super.onResponse(response, params);

                OpenIotService service = (OpenIotService)params[1];
                service.userProjectsResponse(response, params);
            }
        }
                .setToken(this.token)
                .setHttpMethod("GET")
                .send(this.baseUrl + "/projects", this);
    }

    private void userProjectsResponse(String response, Object... params)
    {
        ArrayList<Project> result = new ArrayList<Project>();
        try {
            JSONObject jsonResponse = new JSONObject(response);

            JSONArray jsonArray = jsonResponse.getJSONArray("userProjects");
            for (int i = 0; i < jsonArray.length(); i++) {
                JSONObject jsonProject = jsonArray.getJSONObject(i);

                Project p = new Project();
                p.LoadFromJSON(jsonProject);
                result.add(p);
            }

            this.onUserProjectsResponse((Project[]) result.toArray(new Project[result.size()]), params);

        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    public void onUserProjectsResponse(Project[] projects, Object... params)
    {

    }



    public void requestUserProject(String projectId, Object... params)
    {
        new WebRequest(){
            @Override
            public void onResponse(String response, Object... params) {
                super.onResponse(response, params);

                OpenIotService service = (OpenIotService)params[1];
                service.userProjectResponse(response, params);
            }
        }
                .setToken(this.token)
                .setHttpMethod("GET")
                .send(this.baseUrl + "/project/" + projectId, this);
    }

    private void userProjectResponse(String response, Object... params)
    {
        try{
        JSONObject jsonResponse = new JSONObject(response);

        JSONObject projectJson = jsonResponse.getJSONObject("project");

        PresetsCollection resultPresets = new PresetsCollection();
        JSONArray presetsJson = jsonResponse.getJSONArray("projectPresets");
        for (int i = 0; i < presetsJson.length(); i++)
        {
            Preset preset = new Preset();
            preset.LoadFromJSON(presetsJson.getJSONObject(i));
            resultPresets.add(preset);
        }

        Project resultProject = new Project();
        if (resultProject.LoadFromJSON(projectJson))
            this.onUserProjectResponse(resultProject, resultPresets, params);

        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    public void onUserProjectResponse(Project project, PresetsCollection presets, Object... params)
    {

    }



    public void requestProjectPresets(String projectId)
    {
        new WebRequest(){
            @Override
            public void onResponse(String response, Object... params) {
                super.onResponse(response, params);

                OpenIotService service = (OpenIotService)params[1];
                service.projectPresetsResponse(response, params);
            }
        }
            .setToken(this.token)
            .setHttpMethod("GET")
            .send(this.baseUrl + "/project/" + projectId + "/presets", this);
    }

    private void projectPresetsResponse(String response, Object... params)
    {
        PresetsCollection result = new PresetsCollection();
        try {
            JSONObject jsonResponse = new JSONObject(response);

            JSONArray jsonArray = jsonResponse.getJSONArray("projectPresets");
            for (int i = 0; i < jsonArray.length(); i++) {
                JSONObject jsonPreset = jsonArray.getJSONObject(i);

                Preset p = new Preset();
                p.LoadFromJSON(jsonPreset);
                result.add(p);
            }

            this.onProjectPresetsResponse(result, params);

        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    public void onProjectPresetsResponse(PresetsCollection presets, Object... params)
    {

    }



    public void requestSaveProjectPreset(String projectId, Preset preset)
    {
        new WebRequest(){
            @Override
            public void onResponse(String response, Object... params) {
                super.onResponse(response, params);

                OpenIotService service = (OpenIotService)params[1];
                service.saveProjectPresetResponse(response, params);
            }
        }
                .setToken(this.token)
                .setRequestBody(preset.SaveToJSON().toString())
                .setHttpMethod("PUT")
                .send(this.baseUrl + "/project/" + projectId + "/presets", this);
    }

    private void saveProjectPresetResponse(String response, Object... params)
    {
        try{
            JSONObject jsonResponse = new JSONObject(response);

            Preset result = new Preset();
            if (result.LoadFromJSON(jsonResponse.getJSONObject("projectPreset")))
                this.onSaveProjectPresetResponse(result, params);

        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    public void onSaveProjectPresetResponse(Preset preset, Object... params)
    {

    }



    public void requestUpdateProjectPreset(Preset preset)
    {
        new WebRequest(){
            @Override
            public void onResponse(String response, Object... params) {
                super.onResponse(response, params);

                OpenIotService service = (OpenIotService)params[1];
                service.updateProjectPresetResponse(response, params);
            }
        }
                .setToken(this.token)
                .setRequestBody(preset.SaveToJSON().toString())
                .setHttpMethod("POST")
                .send(this.baseUrl + "/project/" + preset.projectId + "/preset/" + preset.projectPresetId, this);
    }

    private void updateProjectPresetResponse(String response, Object... params)
    {
        try{
            JSONObject jsonResponse = new JSONObject(response);

            Preset result = new Preset();
            if (result.LoadFromJSON(jsonResponse.getJSONObject("projectPreset")))
                this.onUpdateProjectPresetResponse(result, params);

        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    public void onUpdateProjectPresetResponse(Preset preset, Object... params)
    {

    }



    public void requestDeleteProjectPreset(String projectId, final String presetId)
    {
        new WebRequest(){
            @Override
            public void onResponse(String response, Object... params) {
                super.onResponse(response, params);

                OpenIotService service = (OpenIotService)params[1];
                service.deleteProjectPresetResponse(response, presetId, params);
            }
        }
                .setToken(this.token)
                .setHttpMethod("DELETE")
                .send(this.baseUrl + "/project/" + projectId + "/preset/" + presetId, this);
    }

    private void deleteProjectPresetResponse(String response, String presetId, Object... params)
    {
        try{
            JSONObject jsonResponse = new JSONObject(response);

            this.onDeleteProjectPresetResponse(presetId, params);

        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    public void onDeleteProjectPresetResponse(String presetId, Object... params)
    {

    }



    public void requestUpdateDevice(final Device device)
    {
        new WebRequest(){
            @Override
            public void onResponse(String response, Object... params) {
                super.onResponse(response, params);

                OpenIotService service = (OpenIotService)params[1];
                service.requestUpdateDeviceResponse(response, device, params);
            }
        }
                .setToken(this.token)
                .setRequestBody(device.SaveToJSON().toString())
                .setHttpMethod("POST")
                .send(this.baseUrl + "/user/devices", this);
    }

    private void requestUpdateDeviceResponse(String response, Device device, Object... params)
    {
        try{
            JSONObject jsonResponse = new JSONObject(response);

            this.onRequestUpdateDeviceResponse(device, params);

        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    public void onRequestUpdateDeviceResponse(Device device, Object... params)
    {

    }

}

File: Windows/OpenIoT.App/OpenIoT.Lib/Web/Api/OpenIoTService.cs
using OpenIoT.Lib.Web.Models;
using OpenIoT.Lib.Web.Network;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Threading.Tasks;

namespace OpenIoT.Lib.Web.Api
{
    public class OpenIoTService
    {
        public string? Token { get; set; }

        public string BaseUrl { get; set; }
        
        private JsonSerializerOptions serializationOptions;

        public OpenIoTService()
        {
            this.Token = null;

            this.BaseUrl = "http://api.iot.cafe";

            this.serializationOptions = new JsonSerializerOptions()
            {
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
                WriteIndented = true,
                PropertyNameCaseInsensitive = true,
                IncludeFields = false,
                NumberHandling = System.Text.Json.Serialization.JsonNumberHandling.AllowReadingFromString
            };
        }

        public async Task<UserLogin?> RequestUserLogin(String email, String password)
        {
            string response = await new NetworkRequest()
            {
                RequestBody = "{\"email\":\"" + email + "\",\"password\":\"" + password + "\"}",
                HttpMethod = "POST"
            }.SendRequestAsync(this.BaseUrl + "/user/login");

            UserLogin? userLogin = this.DeserializeJson<UserLogin>(response);

            this.Token = userLogin?.Token;

            return userLogin;
        }

        public async Task<IEnumerable<Project>> RequestUserProjects()
        {
            string response = await new NetworkRequest()
            {
                BearerToken = this.Token,
                HttpMethod = "GET"
            }.SendRequestAsync(this.BaseUrl + "/projects");

            using (JsonDocument jdoc = JsonDocument.Parse(response))
            {
                IEnumerable<Project>? projectsList = this.DeserializeJson<List<Project>>(jdoc.RootElement.GetProperty("userProjects"));

                return projectsList;
            }
        }

        public async Task<Tuple<Project, PresetsCollection>> RequestUserProject(string projectId)
        {
            string response = await new NetworkRequest()
            {
                BearerToken = this.Token,
                HttpMethod = "GET"
            }.SendRequestAsync(this.BaseUrl + "/project/" + projectId);

            using (JsonDocument jdoc = JsonDocument.Parse(response))
            {
                Project? project = jdoc.RootElement.GetProperty("project").Deserialize<Project>(this.serializationOptions);
                PresetsCollection? projectPresets = this.DeserializeJson<PresetsCollection>(jdoc.RootElement.GetProperty("projectPresets"));

                return new Tuple<Project, PresetsCollection>(project, projectPresets);
            }
        }

        public async Task<PresetsCollection> RequestProjectPresets(string projectId)
        {
            string response = await new NetworkRequest()
            {
                BearerToken = this.Token,
                HttpMethod = "GET"
            }.SendRequestAsync(this.BaseUrl + "/project/" + projectId + "/presets");

            using (JsonDocument jdoc = JsonDocument.Parse(response))
            {
                PresetsCollection? presets = this.DeserializeJson<PresetsCollection>(jdoc.RootElement.GetProperty("projectPresets"));

                return presets;
            }
        }

        public async Task<Preset> RequestSaveProjectPreset(string projectId, Preset preset)
        {
            string response = await new NetworkRequest()
            {
                BearerToken = this.Token,
                HttpMethod = "POST",
                RequestBody = this.SerializeObject(preset)
            }.SendRequestAsync(this.BaseUrl + "/project/" + projectId + "/presets");

            using (JsonDocument jdoc = JsonDocument.Parse(response))
            {
                Preset? projectPreset = this.DeserializeJson<Preset>(jdoc.RootElement.GetProperty("projectPreset"));

                return projectPreset;
            }
        }

        public async Task<Preset> RequestUpdateProjectPreset(Preset preset)
        {
            string response = await new NetworkRequest()
            {
                BearerToken = this.Token,
                HttpMethod = "PUT",
                RequestBody = this.SerializeObject(preset)
            }.SendRequestAsync(this.BaseUrl + "/project/" + preset.ProjectId + "/preset/" + preset.ProjectPresetId);

            using (JsonDocument jdoc = JsonDocument.Parse(response))
            {
                Preset? projectPreset = this.DeserializeJson<Preset>(jdoc.RootElement.GetProperty("projectPreset"));

                return projectPreset;
            }
        }

        public async Task<bool> RequestDeleteProjectPreset(Preset preset)
        {
            string response = await new NetworkRequest()
            {
                BearerToken = this.Token,
                HttpMethod = "DELETE",
                RequestBody = this.SerializeObject(preset)
            }.SendRequestAsync(this.BaseUrl + "/project/" + preset.ProjectId + "/preset/" + preset.ProjectPresetId);

            return this.IsResponseSuccess(response);
        }

        public string SerializeObject(object obj)
        {
            return JsonSerializer.Serialize(obj, this.serializationOptions);
        }

        public T DeserializeJson<T>(JsonElement json)
        {
            return json.Deserialize<T>(this.serializationOptions);
        }

        public T DeserializeJson<T>(string json)
        {
            return JsonSerializer.Deserialize<T>(json, this.serializationOptions);
        }

        public bool IsResponseSuccess(string response)
        {
            using (JsonDocument jdoc = JsonDocument.Parse(response))
            {
                int responseCode;
                if (!jdoc.RootElement.GetProperty("responseCode").TryGetInt32(out responseCode))
                    return false;

                return responseCode == 1;
            }
        }
    }
}
